horovod/run/gloo_run.py

```diff
@@ -234,79 +184,122 @@ def _exec_command(command, alloc_info, event):
+    run_command = get_run_command(command, server_ip, nics, global_rendezv_port)
+
+    slot_info_to_command = _slot_info_to_command_fn(run_command, env)
+    event = register_shutdown_event()
+    args_list = [[slot_info_to_command(slot_info), slot_info, [event]]
+                 for slot_info in host_alloc_plan]
+
     # If an error occurs in one thread, entire process will be terminated.
     # Otherwise, threads will keep running.
-    res = threads.execute_function_multithreaded(exec_command, args_list, block_until_all_done=True)
+    res = threads.execute_function_multithreaded(exec_command,
+                                                 args_list,
+                                                 block_until_all_done=True)
 
     for name, value in sorted(res.items(), key=lambda item: item[1][1]):
         exit_code, timestamp = value
         if exit_code != 0:
-            raise RuntimeError('Gloo job detected that one or more processes exited with non-zero '
+            raise RuntimeError('Horovod detected that one or more processes exited with non-zero '
                                'status, thus causing the job to be terminated. The first process '
                                'to do so was:\nProcess name: {name}\nExit code: {code}\n'
                                .format(name=name, code=exit_code))
 
 
-def gloo_run(settings, remote_host_names, nics, env, server_ip, command):
+def gloo_run(settings, nics, env, server_ip, command):
     # Each thread will use ssh command to launch the job on each remote host. If an
     # error occurs in one thread, entire process will be terminated. Otherwise,
     # threads will keep running and ssh session.
-    exec_command = _exec_command_fn(settings, remote_host_names)
+    exec_command = _exec_command_fn(settings)
     launch_gloo(command, exec_command, settings, nics, env, server_ip)
+
+
+def gloo_run_elastic(settings, env, command):
+    # Make the output directory if it does not exist
+    if settings.output_filename:
+        _mkdir_p(settings.output_filename)
+
+    rendezvous = RendezvousServer(settings.verbose)
+    driver = ElasticDriver(rendezvous, settings.discovery,
+                           settings.min_np, settings.max_np,
+                           verbose=settings.verbose)
+
+    handler = create_rendezvous_handler(driver)
+    global_rendezv_port = rendezvous.start_server(handler)
+    driver.wait_for_available_hosts(settings.num_proc)
+
+    nics = driver_service.get_common_interfaces(settings, driver.get_available_hosts())
+    server_ip = network.get_driver_ip(nics)
+
+    exec_command = _exec_command_fn(settings)
+    event = register_shutdown_event()
+    run_command = get_run_command(command, server_ip, nics, global_rendezv_port, elastic=True)
+    create_worker = _create_elastic_worker_fn(exec_command, run_command, env, event)
+
+    driver.start(settings.num_proc, create_worker)
+    res = driver.get_results()
+    driver.stop()
+
```

Comment:
I think we need to shut down the rendezvous server here:

        rendezvous.stop_server()

Reply:
Good catch!